svg/types/scripted/SVGGeometryElement.getTotalLength-02.svg

```diff
@@ -0,0 +1,35 @@
+<svg xmlns="http://www.w3.org/2000/svg"
+     xmlns:xlink="http://www.w3.org/1999/xlink"
+     xmlns:html="http://www.w3.org/1999/xhtml">
+  <title>When SVGGeometryElement.getTotalLength is called with an element that is not in the document, throw exception</title>
+  <html:link rel="help" href="https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__getTotalLength"/>
+  <html:script src="/resources/testharness.js"/>
+  <html:script src="/resources/testharnessreport.js"/>
+  <script><![CDATA[
+  test(function() {
+    var pathElement = document.createElementNS("http://www.w3.org/2000/svg", "path");
+    pathElement.setAttribute("d", 'M0,20 L400,20 L640,20');
+    assert_approx_equals(pathElement.getTotalLength(), 640, 1e-5);
+  }, document.title + " with SVGPathElement");
+
+  test(function() {
+    var pathElement = document.createElementNS("http://www.w3.org/2000/svg", "path");
+    assert_throws_dom("InvalidStateError", function() { pathElement.getTotalLength() });
+  }, document.title + " with SVGPathElement (no 'd' attribute)");
+
+  test(function() {
+    var rectElement = document.createElementNS("http://www.w3.org/2000/svg", "rect");
+    rectElement.setAttribute("rx", 0);
+    rectElement.setAttribute("ry", 0);
+    rectElement.setAttribute("width", 200);
+    rectElement.setAttribute("height", 300);
+    assert_throws_dom("InvalidStateError", function() { rectElement.getTotalLength(); });
```

Comment:
In the firefox, does the case return 0 as getBBox?

Reply:
With above test, I called rectElement.getPointAtLength(0) in Firefox. But Firefox throws the exception.
In this case, i thought Firefox is not able to compute the length. Is it right?

Reply:
Looks like, despite our best efforts we've broken that case when we converted to mapping attributes into styles. I'll look into fixing that.

Reply:
> I don't see why we should mandate throwing an invalid state error here.
The reason that it throws an invalid state here is to check that current element is a non-rendered element, and the UA is not able to compute the total length of the path[1]. And if <rect>, <circle> and so on are not in active document, Firefox returns '0' in case of a test[2]. Following a sepc that the UA is not able to compute the total length of the path, i guess that we should throw an invalid state error.
And also I think that <rect>, <circle> and so on that are not in active document is the same as a case that an element has "display:none".

[1] https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__getTotalLength
[2] ![getTotalLength](https://user-images.githubusercontent.com/5849168/93282122-9fa46780-f808-11ea-9230-842fd9dd1bf4.png)

Reply:
Yes Firefox has had a bug since Firefox 69 I think that needs fixing. Once that is fixed it will return the correct value, it won't throw.

Yes, not in an active document is pretty much the same as display: none. In both cases there's no CSS applied to the element. That's the current problem in Firefox that we don't fall back to the element's attributes and try to use them.

Reply:
I have a fix for the Firefox 69 regression. https://phabricator.services.mozilla.com/D90440 when this lands Firefox will return 1000 again.

Reply:
I have a question. Is it possible to get the computed value of 'element' that is not in active?

Reply:
Well the attribute has a value, you can get that using getAttribute so you don't need any computed values here. We've an attribute, we map it to style but if that mapping fails we can just read the attribute value and use that if we can.

Reply:
If the svg element has CSS style, does the gecko map it to style? 
For example, style='d:path(...)'

Reply:
gecko does not currently support path's d atttibute as a style at all. We do support rect x, y, width, height etc as styles (length attributes in most cases). I'm fixing gecko to use attributes as a fallback if style mapping fails.

Reply:
I've fixed the gecko regression. You should be able to try it in a nightly within a day or two.

Reply:
@longsonr
If it is possible, Could you look at your patch?

Reply:
If gecko will support CSS style for svg such as style='r:10' and 'd:path(...)', would gecko need computed style?

Reply:
Could I look at my patch in what regard? If you want to see the patch you should be able to via the phabricator link or via the bugzilla bug I cc'd you on. What happens (post my patch) in gecko is that if there is a style available we'll use it, if not we'll use the attribute value directly. If the value is auto or calc or something then we won't be able to make any sense of it as a value as we need the style system for that. If it's 100 then we're OK and we can proceed without the style system.

Reply:
I am just planning to refer to gecko's codes to make my patch. First, in this case also make a patch with the attribute value directly.
And Do you think we should return 0 below case(SVGGeometryElement.getTotalLength-04.svg) that elements with initial values?

Reply:
Certainly for circle and rect and ellipse returning 0 would seem to be the right thing to do. For polygons, polylines and paths it's less clear what we should do but I expect that not throwing and returning 0 would be more popular with content authors than throwing.